crates/net/network/src/transactions/validation.rs

```diff
@@ -0,0 +1,364 @@
+#[derive(Debug, Default, Deref, DerefMut)]
+pub struct Announcement68Filter<N = LayerOne>(N);
+
+/// L1 Network.
+#[derive(Debug, Display, Default)]
+pub struct LayerOne;
+
+impl ValidateTx68 for LayerOne {
+    fn should_fetch(&self, ty: u8, hash: TxHash, size: usize) -> ValidationOutcome {
+        // 1. checks if tx type is valid value for this network
+        let tx_type = match TxType::try_from(ty) {
+            Ok(ty) => ty,
+            Err(_) => {
+                debug!(target: "net::eth-wire",
+                    ty=ty,
+                    size=size,
+                    hash=%hash,
+                    network=%Self,
+                    "invalid tx type in eth68 announcement"
+                );
+
+                return ValidationOutcome::ReportPeer
+            }
+        };
+        // 2. checks if tx's encoded length is within limits for this network
+        if size > tx_type.max_encoded_tx_length() || size < tx_type.min_encoded_tx_length() {
+            // todo: min valid tx encoded length on layer 1? signature length?
+            debug!(target: "net::eth-wire",
+                ty=ty,
+                size=size,
+                hash=%hash,
+                max_encoded_tx_length=tx_type.max_encoded_tx_length(),
+                network=%Self,
+                "invalid tx size in eth68 announcement"
+            );
+
+            return ValidationOutcome::Ignore
+        }
+
+        ValidationOutcome::Fetch
+    }
+}
+
+impl FilterAnnouncement68 for LayerOne
+where
+    Self: ValidateTx68,
+{
+    fn filter_valid_entries(&self, msg: &mut NewPooledTransactionHashes68) -> FilterOutcome {
+        trace!(target: "net::tx::validation",
+            types=?msg.types,
+            sizes=?msg.sizes,
+            hashes=?msg.hashes,
+            network=%Self,
+            "validating eth68 announcement data.."
+        );
+
+        // 1. checks if the announcement is empty
+        if msg.hashes.is_empty() {
+            debug!(target: "net::tx",
+                network=%Self,
+                "empty announcement"
+            );
+            return FilterOutcome::ReportPeer
+        }
+
+        let mut should_report_peer = false;
+        let mut indices_to_remove = vec![];
+
+        for (i, (&ty, &hash, &size)) in izip!(&msg.types, &msg.hashes, &msg.sizes).enumerate() {
+            match self.should_fetch(ty, hash, size) {
+                ValidationOutcome::Fetch => (),
+                ValidationOutcome::Ignore => indices_to_remove.push(i),
+                ValidationOutcome::ReportPeer => {
+                    indices_to_remove.push(i);
+                    should_report_peer = true;
+                }
+            }
+        }
+
+        for (i, &index) in indices_to_remove.iter().enumerate() {
+            msg.hashes.remove(index - i);
+            msg.types.remove(index - i);
+            msg.sizes.remove(index - i);
+        }
+
+        let mut hashes_sorted = msg.hashes.clone();
+        hashes_sorted.sort();
```

Comment:
I wonder what's more expensive here, sorting or checking duplicates via HashMap inserts

we could also record the indices of duplicated hashes by iterating over them and check hashmap insert and then remove afterwards like above

Reply:
I thought about that too. if we returned a hash map, which would make sense in the calling context, then hash map is better. but, not if we mutate the hashes, sizes and types list in place, then it's just extra work. an issue can be opened to impl this optimisation, think it's not that important and time consuming enough to postpone since changes method signature.

Reply:
I spent some more time thinking about this and remembered another reason for letting it be: https://github.com/paradigmxyz/reth/issues/6148 . if this issue is addressed, instead of passing a `NewPooledTransactionHashes68` to validation, we can pass it a vector of `UnknownTxHash68`s. then we won't the problem of removing entries from hashes, types and sizes lists independently like I describe above, and then it makes sense to use a `HashSet` in this case (`HashMap` not relevant anymore) and return that to the caller context.

...writing about this reminds me probably want to do the duplicate hash validation and empty announcement validation on eth66 announcements too. that can then also return a `HashSet` of the valid hashes. the signature of the filter method would then be to take some `&mut Vec<T>` where T is either `UnknownTxHash68` or `UnknownTxHash66`. `T` can also be an associated type on the filter trait.

the first thing that we would do in the `TransactionsManager::on_new_pooled_transaction_hashes` pipeline is to destruct the announcement, and get a `Vec<TxHash68>` or `Vec<TxHash66>`. the trait method on the pool `retain_unkown` would then also have to be updated to take a `Vec<T>` . alternatively this solution, and handle retaining the hashes for which the check passes in the tx manager instead.
```rust
/// Returns a closure that checks if given transactions is present in the pool.
pub(crate) fn is_tx_known(&self) -> impl FnOnce(TxHash) -> bool + '_ {
    let pool = self.get_pool_data();

    move |tx| !pool.contains(&tx)
}
```

Reply:
@mattsse

Reply:
I wonder if we could simplify this by just collecting into a hashset / map, and setting `should_report_peer` if the size of the map is less than the original size of the announcement

Reply:
yeah it's faster than sort, I believe so.

memory allocation wise though, if the hash map/set is used instead, then I have to 1) allocate the hash map 2) from the hash map allocate new vector(s) for hashes (and sizes and types in eth66), to replace those in the messages. the second step is nonsense. so it would be good to return the hash map/set.

the thing is here, what type of collection can I return? hash map and hash set don't share some collection trait afaik. eth68 needs a `HashMap<TxHash, (u8, usize)` and eth66 just has `HashSet<TxHash>`. a `HashMap<TxHash, Option<(u8, usize)>` can be returned, sure. not very nice type system, but placeholder until https://github.com/paradigmxyz/reth/issues/6148 is addressed.

Reply:
I see, makes sense